### PR TITLE
Add more Prometheus metrics

### PR DESCRIPTION
**Pull request by ronensc**
This PR adds the Prometheus metrics defined in #2650

### CODE REVIEW

File: vllm/engine/metrics.py [hunk 1]
```diff
@@ -44,6 +47,22 @@ def add_global_metrics_labels(**kwargs):
...
+    ],
+)
+histogram_request_generation_tokens = Histogram(
+    "vllm:request_generation_tokens",
+    "Number of generation tokens processed.",
+    buckets=[
+        1, 2, 5, 10, 20, 50, 100, 200, 500, 1_000, 2_000, 5_000, 10_000,
+        20_000, 50_000, 100_000
+    ],
+)
```
Comment (ronensc) [inline vllm/engine/metrics.py:65]: Should we consider the value of  [`max_model_len`](https://github.com/vllm-project/vllm/blob/c9b45adeeb0e5b2f597d1687e0b8f24167602395/vllm/config.py#L50-L51) when setting the buckets? IIUC, `max_model_len` is the upper bound.

**Comment by ronensc** [comment]
@rib-2, I highly value your opinion. Would you please review my pull request?

**Comment by simon-mo** [comment]
@robertgshaw2-neuralmagic

**Comment by ronensc** [comment]
@simon-mo Could you please review this PR?

File: vllm/engine/metrics.py [hunk 2]
```diff
@@ -81,6 +97,18 @@ def __init__(self, labelnames: List[str]):
...
+            documentation="Histogram of the max_tokens request parameter.",
+            labelnames=labelnames,
+            buckets=build_1_2_5_buckets(max_model_len),
+        )
+        self.histogram_request_n = Histogram(
+            name="vllm:request_params_n",
+            documentation="Histogram of the n request parameter.",
+            labelnames=labelnames,
+            buckets=[1, 2, 5, 10, 20],
+        )
```
Comment (hmellor) [inline vllm/engine/metrics.py:111]: Maybe these should go in an `Info`, like the `cache_config`?

File: vllm/engine/metrics.py [hunk 3]
```diff
@@ -60,6 +60,22 @@ def __init__(self, labelnames: List[str]):
...
+            documentation="Number of prefill tokens processed.",
+            labelnames=labelnames,
+            buckets=build_1_2_5_buckets(max_model_len),
+        )
+        self.histogram_request_generation_tokens = Histogram(
+            name="vllm:request_generation_tokens",
+            documentation="Number of generation tokens processed.",
+            labelnames=labelnames,
+            buckets=build_1_2_5_buckets(max_model_len),
+        )
```
[the remark this answers is not in the document]
Comment (ronensc) [reply vllm/engine/metrics.py:78]: Thanks for your feedback!
You are right. But, wouldn't it be beneficial to have in addition histograms depicting the distribution of prompt length and generation length?

File: vllm/engine/metrics.py [hunk 4]
```diff
@@ -60,6 +60,22 @@ def __init__(self, labelnames: List[str]):
             name="vllm:generation_tokens_total",
             documentation="Number of generation tokens processed.",
             labelnames=labelnames)
+        self.counter_request_success = Counter(
+            name="vllm:request_success",
+            documentation="Count of successfully processed requests.",
+            labelnames=labelnames)
```
[the remark this answers is not in the document]
Comment (ronensc) [reply vllm/engine/metrics.py:66]: A quick option to add http related metrics would be to use [`prometheus-fastapi-instrumentator`](https://github.com/trallnag/prometheus-fastapi-instrumentator/tree/master).

This involves installing the package:
```pip install prometheus-fastapi-instrumentator```

Then adding the following 2 lines after the app creation:
https://github.com/vllm-project/vllm/blob/8af890a865bf9f744d7d9bd5515558b42224c744/vllm/entrypoints/openai/api_server.py#L47-L48
```
from prometheus_fastapi_instrumentator import Instrumentator
Instrumentator().instrument(app).expose(app)
```

This will add the following metrics:

| Metric Name                    | Type     | Description                                                                                                     |
|-------------------------------|----------|-----------------------------------------------------------------------------------------------------------------|
| http_requests_total            | counter  | Total number of requests by method, status, and handler.                                                        |
| http_request_size_bytes        | summary  | Content length of incoming requests by handler. Only value of header is respected. Otherwise ignored.         |
| http_response_size_bytes       | summary  | Content length of outgoing responses by handler. Only value of header is respected. Otherwise ignored.         |
| http_request_duration_highr_seconds | histogram | Latency with many buckets but no API specific labels. Made for more accurate percentile calculations.       |
| http_request_duration_seconds  | histogram | Latency with only a few buckets by handler. Made to be only used if aggregation by handler is important.     |



Should I add it to the PR?

File: vllm/engine/metrics.py (hunk 2 above)
Comment (ronensc) [reply vllm/engine/metrics.py:111]: IIUC, `Info` is intended to collect constant metrics such as configuration values. `vllm:request_params_max_tokens` and `vllm:request_params_n` are intended to collect the values of the `max_tokens` and `n` arguments in each request, which may vary with each request.
For instance, in the following request, `max_tokens=7` and `n=3` will be collected.
```
curl http://localhost:8000/v1/completions \
    -H "Content-Type: application/json" \
    -d '{
        "model": "facebook/opt-125m",
        "prompt": "San Francisco is a",
        "max_tokens": 7,
        "best_of": 5,
        "n": 3,
        "use_beam_search": "true",
        "temperature": 0
    }'
```

File: vllm/engine/metrics.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (hmellor) [reply vllm/engine/metrics.py:66]: I like this solution, it saves us from reinventing the wheel in vLLM

File: vllm/engine/metrics.py (hunk 2 above)
Comment (hmellor) [reply vllm/engine/metrics.py:111]: Good point.

Although, I'm not convinced that it's a useful datapoint to collect (I am open to being convinced though!). i.e. what do we, as the vLLM provider, learn from the token limit the user sets?

Also, in the vast majority of use cases a user will statically set their sampling parameters in their application and then never touch them again.

File: vllm/engine/metrics.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (hmellor) [reply vllm/engine/metrics.py:78]: Since this metric doesn't actually introduce any overhead (because the data from `vllm:x_tokens_total` is reused, these two are probably fine. It would be interesting to know how big the prompts the users were providing are.

[the remark this answers is not in the document]
Comment (ronensc) [reply vllm/engine/metrics.py:80]: Exactly! I suggest to deprecate the 2 `vllm:x_tokens_total` metrics as they will be included as part of the Histogram metrics this PR adds.

File: vllm/engine/metrics.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ronensc) [reply vllm/engine/metrics.py:68]: done :)

File: vllm/entrypoints/openai/api_server.py [hunk 5]
```diff
@@ -45,18 +45,15 @@ async def _force_log():
...
 def parse_args():
     parser = make_arg_parser()
     return parser.parse_args()
 
 
-# Add prometheus asgi middleware to route /metrics requests
-metrics_app = make_asgi_app()
-app.mount("/metrics", metrics_app)
-
-
```
Comment (hmellor) [inline vllm/entrypoints/openai/api_server.py:59]: This is how the metrics defined in `vllm/engine/metrics.py` are exposed. It can't be removed.

File: vllm/engine/metrics.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (hmellor) [reply vllm/engine/metrics.py:68]: Nice, if we're going to be using `prometheus-fasapi-instrumentator` then this implementation should be removed.

File: vllm/engine/metrics.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (hmellor) [reply vllm/engine/metrics.py:80]: I think we should keep these metrics, because a developer may not want to have to aggregate histogram data in order to get the same effect of `vllm:x_tokens_total`

[the remark this answers is not in the document]
Comment (ronensc) [reply vllm/engine/metrics.py:80]: Prometheus histograms have this nice feature where in addition to the bucket counters, they include 2 additional counters 
suffixed with `_sum` and `_count`.

`_count` is incremented by 1 on every observe, and `_sum` is incremented by the value of the observation.

Therfore, `vllm:prompt_tokens_total` is equivalent to `vllm:request_prompt_tokens_sum`,
and `vllm:generation_tokens_total` is equivalent to `vllm:request_generation_tokens_sum`

Source:
https://www.robustperception.io/how-does-a-prometheus-histogram-work/

File: vllm/engine/metrics.py (hunk 2 above)
Comment (ronensc) [reply vllm/engine/metrics.py:113]: I see your point. Having a histogram of the token limit set by the user might indeed be redundant, particularly considering we have a histogram of the actual number of generated tokens per request (`vllm:request_generation_tokens_sum`). I'll remove it.


Regarding `n` and `best_of`, I realized I overlooked adding a metric for collecting `best_of`. Just to clarify, `best_of` determines the width of the beam search, while `n` specifies how many "top beams" to return (`n` <= `best_of`). AFAIU, larger values of `best_of` can significantly impact the engine, as the batch will be dominated by sequences from a few requests.

It might be insightful to compare the histograms of both `n` and `best_of`. A significant deviation between them could suggest that the engine is processing a substantial number of tokens that users aren't actually consuming.

Please let me know your thoughts on this.

File: vllm/entrypoints/openai/api_server.py (hunk 5 above)
Comment (ronensc) [reply vllm/entrypoints/openai/api_server.py:59]: I replaced it with the `expose()` method of `prometheus-fastapi-instrumentator` which also exposes a `/metric` endpoint.
https://github.com/vllm-project/vllm/pull/2764/files#diff-38318677b76349044192bf70161371c88fb2818b85279d8fc7f2c041d83a9544R48-R49

I noticed it also solves the `/metrics/` redirection [issue](https://github.com/prometheus/client_python/issues/1016).
Which of the exposing methods should we use?

File: vllm/engine/metrics.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (hmellor) [reply vllm/engine/metrics.py:80]: Oh I see, thanks for explaining. In that case you could move the `vllm:x_tokens_total` metrics into the `# Legacy metrics` section.

Although I think there might be some objection to changing metrics that people are already using in dashboards.

cc @simon-mo @Yard1

File: vllm/engine/metrics.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (hmellor) [reply vllm/engine/metrics.py:68]: The code highlighted in the original comment
```python
        self.counter_request_success = Counter(
            name="vllm:request_success",
            documentation="Count of successfully processed requests.",
            labelnames=labelnames)
```
can be removed if we are getting these metrics from `prometheus-fastapi-instrumentator` instead.

File: vllm/entrypoints/openai/api_server.py (hunk 5 above)
Comment (hmellor) [reply vllm/entrypoints/openai/api_server.py:59]: > I replaced it with the expose() method of prometheus-fastapi-instrumentator which also exposes a /metric endpoint.

While this does expose a `/metrics` endpoint, none of the vLLM metrics will be in it because they come from `make_asgi_app()`, right?

Have you confirmed that `/metrics` still contains vLLM metrics with this code removed?

File: vllm/engine/metrics.py (hunk 2 above)
Comment (hmellor) [reply vllm/engine/metrics.py]: Knowing the amount of tokens "wasted" in beam search could be interesting to know as a developer of vLLM. i.e. if beam search is being used extensively and many tokens are being "wasted", it signals that we need to optimise beam search if we can.

@simon-mo @Yard1 what do you think about this?

File: vllm/engine/metrics.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (simon-mo) [reply vllm/engine/metrics.py:80]: from my point of view, it's fine to duplicate metrics for backward compatibility reason.

**Review by simon-mo — COMMENTED** [review]
deferring to @hmellor for approval (thanks for the detail review), please let me know when it's a good time merge.

File: vllm/entrypoints/openai/api_server.py (hunk 5 above)
Comment (ronensc) [reply vllm/entrypoints/openai/api_server.py:59]: "Yes, I've verified that both approaches expose all metrics. The only discrepancy I've noticed is that `expose()` from `prometheus-fastapi-instrumentator` exposes metrics on `/metrics`, whereas `make_asgi_app()` exposes them on `/metrtics/`. However, I'll revert to using the `make_asgi_app()` approach. I find the other method somewhat hacky, as it involves the `prometheus-fastapi-instrumentator` middleware handling the metrics endpoint. This could look weird if multiple middlewares are in use."

File: vllm/engine/metrics.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ronensc) [reply vllm/engine/metrics.py:68]: Thank you for clarifying. I believe that `vllm:request_success` remains valuable. It includes a `finished_reason` label, which allows for counting requests based on their finished reason — either `stop` if the sequence ends with an EOS token, or `length` if the sequence length reaches either `scheduler_config.max_model_len` or `sampling_params.max_tokens`. I'm open to adjusting its name and description to make it more indicative.

File: vllm/engine/metrics.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ronensc) [reply vllm/engine/metrics.py:80]: Sure, I'll relocate these metrics to the legacy section. Perhaps in the future, when we're able to make breaking changes, we can consider removing them.

File: vllm/engine/metrics.py [hunk 6]
```diff
@@ -165,6 +216,26 @@ def _log_prometheus(self, stats: Stats) -> None:
         self.metrics.counter_generation_tokens.labels(**self.labels).inc(
             stats.num_generation_tokens)
 
+        # Add to request counters.
+        for finished_reason, count in stats.finished_reason_counter.items():
+            self.metrics.counter_request_success.labels({
+                **self.labels,
+                "finished_reason":
+                finished_reason,
```
Comment (hmellor) [inline vllm/engine/metrics.py:224]: Does this not raise an error because `finished_reason` was not specified when this metric was constructed?

File: vllm/engine/metrics.py [hunk 7]
```diff
@@ -82,6 +99,12 @@ def __init__(self, labelnames: List[str]):
             documentation="Histogram of end to end request latency in seconds.",
             labelnames=labelnames,
             buckets=[1.0, 2.5, 5.0, 10.0, 15.0, 20.0, 30.0, 40.0, 50.0, 60.0])
+        self.histogram_request_n = Histogram(
+            name="vllm:request_params_n",
+            documentation="Histogram of the n request parameter.",
+            labelnames=labelnames,
+            buckets=[1, 2, 5, 10, 20],
+        )
```
Comment (hmellor) [inline vllm/engine/metrics.py:107]: If we keep what's currently called `vllm:request_success`, could this be included as another `label` and the histogram could be constructed on the Grafana side?

File: vllm/engine/metrics.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (hmellor) [reply vllm/engine/metrics.py:68]: What do you think of the idea of renaming this to something like `vllm:request_info` and including `n` and `best_of` as `labels` too? This way we log a single metric from which the user can construct many different visualisations on Grafana by utilising the label filters?

[the remark this answers is not in the document]
Comment (ronensc) [reply vllm/engine/metrics.py:68]: While the idea of combining these metrics may seem appealing at first glance, I believe they should be kept separate for the following reasons:

1. The metrics have different types: `vllm:request_success` is a Counter, while `vllm:request_params_best_of` and `vllm:request_params_n` are Histograms.

2. Aggregating different labels lacks semantic meaning.

3. Although merging `n` and `best_of` into the same histograms might make sense in this case, as they would share the same buckets, we may encounter scenarios where we need to introduce another metric with different bucket requirements.

4. This situation differs from the `Info` metric type, where data is encoded in the label values.

File: vllm/engine/metrics.py (hunk 7 above)
Comment (ronensc) [reply vllm/engine/metrics.py:107]: Great suggestion! I'll incorporate the new metrics into the Grafana dashboard demo.

**Comment by ronensc** [comment]
I've added 3 new panels at the bottom of the Grafana dashboard showcasing the metrics `vllm:request_prompt_tokens_bucket`, `vllm:request_generation_tokens_bucket` and `vllm:request_success_total`.
I haven't included panels for `n` and `best_of` since the current benchmarking script sets a constant value to `best_of` and doesn't set `n` at all.

I ran the benchmark on gpt2 model as I couldn't fit `mistralai/Mistral-7B-v0.1` onto my GPU:
```
# Start vLLM
python3 -m vllm.entrypoints.openai.api_server --model gpt2 --disable-log-requests

# Run benchmarking script
python3 ../../benchmarks/benchmark_serving.py --model gpt2 --tokenizer gpt2 --endpoint /v1/completions --dataset ShareGPT_V3_unfiltered_cleaned_split.json --request-rate 10.0
```

image

**Comment by simon-mo** [comment]
wow cool histogram visuals!

**Comment by robertgshaw2-neuralmagic** [comment]
@ronensc is this ready for review?

File: vllm/engine/llm_engine.py [hunk 8]
```diff
@@ -575,26 +577,51 @@ def _get_stats(self,
...
+            # Number of Tokens
             if prompt_run:
-                num_prompt_tokens = sum(
+                num_prompt_tokens_lst = [
                     len(scheduled_seq_group.seq_group.prompt_token_ids)
                     for scheduled_seq_group in
-                    scheduler_outputs.scheduled_seq_groups)
+                    scheduler_outputs.scheduled_seq_groups
+                ]
+                num_prompt_tokens = sum(num_prompt_tokens_lst)
```
Comment (hmellor) [inline vllm/engine/llm_engine.py:598]: Instead of passing both the list and the summed version, can we just pass the list (under the original name, `num_prompt_tokens` and do the `sum()` in `StatLogger._log_prometheus()`? That way deprecated metrics do not effect `llm_engine.py`.

The same comment goes for `num_generation_tokens` and `num_generation_tokens_lst`.

File: vllm/engine/metrics.py (hunk 7 above)
Comment (hmellor) [reply vllm/engine/metrics.py:101]: This can be resolved now as you've added the Grafana examples

**Review by hmellor — CHANGES_REQUESTED** [review]
It's looking really good now, just a few nits

File: vllm/engine/metrics.py [hunk 9]
```diff
@@ -13,6 +13,8 @@
 
 disable_created_metrics()
 
+LABEL_NAME_FINISHED_REASON = "finished_reason"
```
[the remark this answers is not in the document]
Comment (robertgshaw2-neuralmagic) [reply vllm/engine/metrics.py:16]: +1

File: vllm/engine/llm_engine.py [hunk 10]
```diff
@@ -575,26 +577,51 @@ def _get_stats(self,
...
         num_prompt_tokens = 0
         num_generation_tokens = 0
+        num_prompt_tokens_lst = []
+        num_generation_tokens_lst = []
+        request_n = []
+        request_best_of = []
         time_to_first_tokens = []
         time_per_output_tokens = []
         time_e2e_requests = []
+        finished_reason_counter = CollectionsCounter()
```
[the remark this answers is not in the document]
Comment (robertgshaw2-neuralmagic) [reply vllm/engine/llm_engine.py:587]: Is there any reason this can't just be a list of finished reasons that we process on the `stat_logger` side?

[the remark this answers is not in the document]
Comment (hmellor) [reply vllm/engine/llm_engine.py:587]: I don't think so, and it would make the stats logging code in the engine a bit cleaner

File: vllm/engine/llm_engine.py [hunk 11]
```diff
@@ -575,26 +577,51 @@ def _get_stats(self,
...
         if scheduler_outputs is not None:
             prompt_run = scheduler_outputs.num_prefill_groups > 0
 
-            # Number of Tokens.
+            # Number of Tokens
             if prompt_run:
-                num_prompt_tokens = sum(
+                num_prompt_tokens_lst = [
```
Comment (robertgshaw2-neuralmagic) [inline vllm/engine/llm_engine.py:593]: This is actually an incorrect calculation. 

This `stat` is used to compute the histogram of prompt lengths for each request. Because of chunked prefill, the prompt processing will occur over multiple `steps` of LLMEngine. So this will actually give you the distribution of chunks as opposed to the distribution of prefill lengths.

So what I would suggest is that we log the of the generation and the prompt once the `seq_group` is finished (e.g. in the loop where we loop through the `scheduled_sequence_groups`

This way we know that we are logging the full prefill length and that we are logging one item per iteration

File: vllm/engine/llm_engine.py [hunk 12]
```diff
@@ -609,15 +636,29 @@ def _get_stats(self,
             time_to_first_tokens = time_last_iters if prompt_run else []
             time_per_output_tokens = [] if prompt_run else time_last_iters
 
+            # Finished Requests
```
Comment (robertgshaw2-neuralmagic) [inline vllm/engine/llm_engine.py:639]: Its a bit hard to follow where we have 5-6 loops that loop through the same list of `scheduled_seq_groups`

File: vllm/engine/llm_engine.py [hunk 13]
```diff
@@ -575,26 +577,51 @@ def _get_stats(self,
...
                 num_generation_tokens = scheduler_outputs.num_batched_tokens
+                num_generation_tokens_lst = [
+                    seq.get_output_len() for scheduled_seq_group in
+                    scheduler_outputs.scheduled_seq_groups for seq in
+                    scheduled_seq_group.seq_group.get_finished_seqs()
+                ]
 
-            # Latency Timings.
+            # Sampling Params
+            if prompt_run:
```
Comment (robertgshaw2-neuralmagic) [inline vllm/engine/llm_engine.py:612]: Same issue related to chunked prefill. If our prefill has N chunked, we will create N histogram entires

File: vllm/engine/metrics.py [hunk 14]
```diff
@@ -117,6 +117,16 @@ def __init__(self, labelnames: List[str], max_model_len: int):
...
+        # Deprecated in favor of vllm:request_prompt_tokens_sum
+        self.counter_prompt_tokens = Counter(
+            name="vllm:prompt_tokens_total",
+            documentation="Number of prefill tokens processed.",
+            labelnames=labelnames)
+        # Deprecated in favor of vllm:request_generation_tokens_sum
+        self.counter_generation_tokens = Counter(
+            name="vllm:generation_tokens_total",
+            documentation="Number of generation tokens processed.",
+            labelnames=labelnames)
```
[the remark this answers is not in the document]
Comment (robertgshaw2-neuralmagic) [reply vllm/engine/metrics.py:129]: With my suggestion for how to properly compute the request metric summaries, these will no longer be deprecated

[the remark this answers is not in the document]
Comment (robertgshaw2-neuralmagic) [reply vllm/engine/metrics.py:129]: ( since these will compute the "realtime" throughput )
( and the others will compute the histogram of generated tokens / histogram of prefill tokens for completed requests )

**Review by robertgshaw2-redhat — CHANGES_REQUESTED** [review]
Some of the metrics are incorrect due to chunked prefill. This is because the current code makes the implicit assumption that `prompt_run` will happen once per sequence group, which is not true. This causes us to log the request level data N times for a prompt that is split into N chunks

As a result, request level metrics are better logged when the request completes as opposed to when the request arrives such that we know its data will be logged once and only once

**Comment by robertgshaw2-neuralmagic** [comment]
@ronensc Left some comments. 

I also have been working on a branch for expanding metrics. Im going to make a PR to your branch if thats okay.

File: vllm/engine/llm_engine.py [hunk 15]
```diff
@@ -575,26 +577,51 @@ def _get_stats(self,
...
+        finished_reason_counter = CollectionsCounter()
         if scheduler_outputs is not None:
             prompt_run = scheduler_outputs.num_prefill_groups > 0
 
-            # Number of Tokens.
+            # Number of Tokens
             if prompt_run:
-                num_prompt_tokens = sum(
+                num_prompt_tokens_lst = [
                     len(scheduled_seq_group.seq_group.prompt_token_ids)
```
Comment (robertgshaw2-neuralmagic) [inline vllm/engine/llm_engine.py:594]: Actually now that I think about it, we have a bug on current main for this same reason

We are actually processing `chunk_len` tokens rather then `len(prompt_token_ids`) in this iteration

File: vllm/engine/llm_engine.py [hunk 16]
```diff
@@ -609,15 +636,29 @@ def _get_stats(self,
             time_to_first_tokens = time_last_iters if prompt_run else []
```
Comment (robertgshaw2-neuralmagic) [inline vllm/engine/llm_engine.py:636]: these are also now incorrect b/c of chunked prefill

**Comment by robertgshaw2-neuralmagic** [comment]
https://github.com/ronensc/vllm/pull/1

^ draft PR to fix the issues I identified above + added a couple more metrics

@ronensc if you can finish this off it would be great. Otherwise, I will finish it off over the weekend.

File: requirements-common.txt [hunk 17]
```diff
@@ -10,6 +10,7 @@ fastapi
 uvicorn[standard]
 pydantic >= 2.0  # Required for OpenAI server.
 prometheus_client >= 0.18.0
+prometheus-fastapi-instrumentator
```
[the remark this answers is not in the document]
Comment (ronensc) [reply requirements-common.txt:13]: done.

File: vllm/engine/metrics.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (ronensc) [reply vllm/engine/metrics.py:16]: done.

**Review by ronensc — COMMENTED** [review]
Thank you for all your feedback and the thorough review!
@hmellor, I think I've addressed all your comments.
@robertgshaw2-neuralmagic, Thank you for bringing up the chunked prefill issue and the draft PR. I will address your comments later, and integrate the PR you've opened in my fork with this one.

File: vllm/engine/llm_engine.py [hunk 18]
```diff
@@ -575,26 +577,51 @@ def _get_stats(self,
         # Iteration stats if we have scheduler output.
         num_prompt_tokens = 0
         num_generation_tokens = 0
+        num_prompt_tokens_lst = []
+        num_generation_tokens_lst = []
+        request_n = []
+        request_best_of = []
```
[the remark this answers is not in the document]
Comment (hmellor) [reply vllm/engine/llm_engine.py:580]: Ah in the diff view I hadn't realised this was nested, it's ok to leave as is then

**Comment by robertgshaw2-neuralmagic** [comment]
Thanks @ronensc 

This is a bug on main right now, so will get merged asap once ready to go

**Comment by ronensc** [comment]
I see. I'll do my best to deliver ASAP

**Comment by ronensc** [comment]
1. I've incorporated most of the changes from https://github.com/ronensc/vllm/pull/1 into this PR.
2. I'm not sure the whether [assumption](https://github.com/ronensc/vllm/pull/1/files#diff-9d1cd5050a7ec1e588cae646f3f95ca134cffbd8b41d6655c6a14de70117b869R459-R461) in `maybe_get_last_latency()` and `maybe_set_first_token_time()` is correct.
Both methods are called after [`self.output_processor.process_outputs()`](https://github.com/vllm-project/vllm/blob/5ded719d9cfd09b300562ec1d4df21fb0a4e79a3/vllm/engine/llm_engine.py#L459) (and specifically after `seq_group.update_num_computed_tokens()`). By this point, when the first generation token is ready, it is already added to the sequence, so the state of `seq_group` is changed from `PREFILL` to `DECODE` and `get_num_uncomputed_tokens() == 1`.
For `maybe_set_first_token_time()`, I suggest using the condition `self.get_seqs()[0].get_output_len() == 1` to determine when the first token is generated.
As for `maybe_get_last_latency()`, I suggest using the condition `self.is_prefill()` to check when chunked_prefill is ongoing.
3. I modified [`num_generation_tokens_iter += 1 `](https://github.com/ronensc/vllm/pull/1/files#diff-c89ac25bd066e936e80260d21be63c7d2379cfedc371a9ff288fb5ba02ae1350R613) to `num_generation_tokens_iter += seq_group.num_unfinished_seqs()` to accommodate requests with more than one sequence (like beam search and parallel sampling).
4. I'll postpone adding the additional metrics until we get the current set of metrics right.

Note: I applied the changes from https://github.com/vllm-project/vllm/pull/4150 locally to aid in debugging.

**Comment by robertgshaw2-neuralmagic** [comment]
@ronensc is this ready for review?

**Comment by ronensc** [comment]
In current state of the PR, some of the metrics are still inaccurate in chunked_prefill.
Before addressing the chunked_prefill issue, could we please merge this PR up to commit https://github.com/vllm-project/vllm/pull/2764/commits/5ded719d9cfd09b300562ec1d4df21fb0a4e79a3 (before attempting to solve the chunked_prefill issue)? We can tackle the chunked_prefill problem in a follow-up PR. What do you think?
Also, just a heads-up, I'll be less available in the coming days.

**Review by hmellor — APPROVED** [review]
From my perspective this LGTM.

Once @robertgshaw2-neuralmagic also approves, I'd say this is good to merge.

**Comment by njhill** [comment]
Huge thanks for all the work on this and reviews @ronensc @robertgshaw2-neuralmagic @hmellor

**Comment by robertgshaw2-neuralmagic** [comment]
Im just thinking though

"I'm not sure the whether [assumption](https://github.com/ronensc/vllm/pull/1/files#diff-9d1cd5050a7ec1e588cae646f3f95ca134cffbd8b41d6655c6a14de70117b869R459-R461) in maybe_get_last_latency() and maybe_set_first_token_time() is correct.
Both methods are called after [self.output_processor.process_outputs()](https://github.com/vllm-project/vllm/blob/5ded719d9cfd09b300562ec1d4df21fb0a4e79a3/vllm/engine/llm_engine.py#L459) (and specifically after seq_group.update_num_computed_tokens()). By this point, when the first generation token is ready, it is already added to the sequence, so the state of seq_group is changed from PREFILL to DECODE and get_num_uncomputed_tokens() == 1.
For maybe_set_first_token_time(), I suggest using the condition self.get_seqs()[0].get_output_len() == 1 to determine when the first token is generated.
As for maybe_get_last_latency(), I suggest using the condition self.is_prefill() to check when chunked_prefill is ongoing."

Will merge this weekend

**Comment by robertgshaw2-neuralmagic** [comment]
@njhill had to make a couple changes for correctness due to some subtlety with chunked_prefill

**Review by robertgshaw2-redhat — APPROVED** [review]
